src/Http/WebUtilities/src/FileBufferingWriteStream.cs

```diff
@@ -184,9 +186,31 @@ public async Task DrainBufferAsync(Stream destination, CancellationToken cancell
             // unspooled content. Copy the FileStream content first when available.
             if (FileStream != null)
             {
-                FileStream.Position = 0;
-                await FileStream.CopyToAsync(destination, cancellationToken);
+                // We make a new stream for async reads from disk and async writes to the destination
+                await using var readStream = new FileStream(FileStream.Name, FileMode.Open, FileAccess.Read, FileShare.Delete | FileShare.ReadWrite, bufferSize: 1, useAsync: true);
 
+                await readStream.CopyToAsync(destination, cancellationToken);
+
+                // This is created with delete on close
+                await FileStream.DisposeAsync();
```

Comment:
You don't need to keep the original FileStream/handle open if you use delete-on-close - the read stream can use the file normally as long as it needs (the actual filesystem level deletion happens after all handles are closed). So this could be moved before CopyToAsync.

Reply:
I left the behavior as it was.

Reply:
Ah I see what you're saying, does it matter though?

Reply:
Not sure if it's important, but this would just release the object earlier and in case CopyToAsync throws the behavior would be different. But in that case neither the old or new logic is deterministic, so it probably doesn't matter.